sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,290 @@
+            .setTransformId(outputLocation.getTransformId())
+            .setTimerFamilyId(outputLocation.getTimerFamilyId())
+            .setIsLast(true);
+      } else {
+        bufferedElements
+            .addDataBuilder()
+            .setInstructionId(outputLocation.getInstructionId())
+            .setTransformId(outputLocation.getTransformId())
+            .setIsLast(true);
+      }
+    }
+    outboundObserver.onNext(bufferedElements.build());
+    outputLocations.clear();
+    perEndpointByteCount.clear();
+    perEndpointElementCount.clear();
+  }
+
+  public <T> void acceptData(String pTransformId, T data) throws Exception {
+    accept(LogicalEndpoint.data(processBundleRequestIdSupplier.get(), pTransformId), data);
+  }
+
+  public <T> void acceptTimers(String pTransformId, String timerFamilyId, T timers)
+      throws Exception {
+    accept(
+        LogicalEndpoint.timer(processBundleRequestIdSupplier.get(), pTransformId, timerFamilyId),
+        timers);
+  }
+
+  public <T> void accept(LogicalEndpoint endpoint, T data) throws Exception {
+    if (timeLimit > 0) {
+      checkFlushThreadException();
+    }
+    buffer.compute(
+        endpoint,
+        (e, output) -> {
+          if (output == null) {
+            output = ByteString.newOutput();
+          }
+          int size = output.size();
+          try {
+            ((Coder<T>) outputLocations.get(e)).encode(data, output);
+          } catch (IOException ex) {
+            throw new RuntimeException("Failed to encode data.");
+          }
+          if (output.size() - size == 0) {
+            output.write(0);
+          }
+          final long delta = (long) output.size() - size;
+          byteCounter.getAndAdd(delta);
+          perEndpointByteCount.compute(
+              e, (e1, byteCount) -> byteCount == null ? delta : delta + byteCount);
+          return output;
+        });
+
+    if (byteCounter.get() >= sizeLimit) {
+      flush();
+    }
+  }
+
+  private Elements.Builder convertBufferForTransmission() {
+    Elements.Builder bufferedElements = Elements.newBuilder();
+    for (LogicalEndpoint endpoint : buffer.keySet()) {
+      ByteString.Output output = buffer.remove(endpoint);
+      if (output == null) {
+        continue;
+      }
+      byteCounter.getAndAdd(-1 * output.size());
+      if (endpoint.isTimer()) {
+        bufferedElements
+            .addTimersBuilder()
+            .setInstructionId(endpoint.getInstructionId())
+            .setTransformId(endpoint.getTransformId())
+            .setTimerFamilyId(endpoint.getTimerFamilyId())
+            .setTimers(output.toByteString());
+      } else {
+        bufferedElements
+            .addDataBuilder()
+            .setInstructionId(endpoint.getInstructionId())
+            .setTransformId(endpoint.getTransformId())
+            .setData(output.toByteString());
+      }
+      perEndpointElementCount.compute(endpoint, (e, count) -> count == null ? 1 : count + 1);
+      output.reset();
```

Comment:
Since you removed the output there is no point in resetting it since the next `accept` call will add it again.
```suggestion
```